src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,711 @@
+                    let c2 = narrowed_ctx2.narrow(&BitOpStep::from(i));
+                    async move {
+                        let not_saturated_case =
+                            (one - is_saturated).multiply(bit, c1, record_id).await?;
+                        let just_saturated_case = is_saturated_and_prev_row_not_saturated
+                            .multiply(prev_bit, c2, record_id)
+                            .await?;
+                        Ok::<_, Error>(not_saturated_case + &just_saturated_case)
+                    }
+                }),
+        )
+        .await?,
+    ))
+}
+
+///
+/// This function contains the main logic for the per-user attribution circuit.
+/// Multiple rows of data about a single user are processed in-order from oldest to newest.
+///
+/// Summary:
+/// - Last touch attribution
+///     - Every trigger event which is preceded by a source event is attributed
+///     - Trigger events are attributed to the `breakdown_key` of the most recent preceding source event
+/// - Per user capping
+///     - A cumulative sum of "Attributed Trigger Value" is maintained
+///     - Bitwise addition is used, and a single bit indicates if the sum is "saturated"
+///     - The only available values for "cap" are powers of 2 (i.e. 1, 2, 4, 8, 16, 32, ...)
+///     - Prior to the cumulative sum reaching saturation, attributed trigger values are passed along
+///     - The row which puts the cumulative sum over the cap is "capped" to the delta between the cumulative sum of the last row and the cap
+///     - All subsequent rows contribute zero
+/// - Outputs
+///     - If a user has `N` input rows, they will generate `N-1` output rows. (The first row cannot possibly contribute any value to the output)
+///     - Each output row has two main values:
+///         - `capped_attributed_trigger_value` - the value to contribute to the output (bitwise secret-shared),
+///         - `attributed_breakdown_key` - the breakdown to which this contribution applies (bitwise secret-shared),
+///     - Additional output:
+///         - `did_trigger_get_attributed` - a secret-shared bit indicating if this row corresponds to a trigger event
+///           which was attributed. Might be able to reveal this (after a shuffle and the addition of dummies) to minimize
+///           the amount of processing work that must be done in the Aggregation stage.
+async fn compute_row_with_previous<C, BK, TV>(
+    ctx: C,
+    record_id: RecordId,
+    input_row: &PrfShardedIpaInputRow<BK, TV>,
+    inputs_required_from_previous_row: &mut InputsRequiredFromPrevRow,
+    num_saturating_sum_bits: usize,
+) -> Result<CappedAttributionOutputs, Error>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    BK: GaloisField,
+    TV: GaloisField,
+{
+    let bd_key = BitDecomposed::decompose(BK::BITS, |i| {
+        input_row.breakdown_key.map(|v| Gf2::truncate_from(v[i]))
+    });
```

Comment:
We might want to consider a function for this conversion on GfXXX types.

Reply:
That makes sense.